spec/services/exchange_rates/average_exchange_rates_service_spec.rb

```diff
@@ -0,0 +1,355 @@
+require 'rails_helper'
+
+RSpec.describe ExchangeRates::AverageExchangeRatesService do
+  describe '.call' do
+    subject(:create_average_rates) { described_class.call(force_run) }
+
+    let(:force_run) { false }
+
+    let(:expected_filepath) { "data/exchange_rates/#{test_date.year}/#{test_date.month}/average_csv_#{test_date.year}-#{test_date.month}.csv" }
+    let(:expected_csv) do
+      <<~CSV
+        Country,Unit Of Currency,Currency Code,Sterling value of Currency Unit £,Currency Units per £1
+        Australia,Dollar,AUD,0.5264,1.8997
+        Dubai,Dirham,AED,0.2286,4.3747
+        Eurozone,Euro,EUR,0.8702,1.1492
+        Kazakhstan,Tenge,USD,0.8154,1.2264
+        Kazakhstan,Euro,EUR,0.8702,1.1492
+        Kazakhstan,Tenge,KZT,0.0018,545.0922
+        United States,Dollar,USD,0.8154,1.2264
+      CSV
+    end
+
+    before do
+      travel_to test_date
+      setup_data
+
+      allow(TariffSynchronizer::FileService)
+       .to receive(:write_file)
+       .with(expected_filepath, expected_csv)
+    end
+
+    shared_examples 'creating the average rates' do
+      it 'creates the average rates', :aggregate_failures do
+        create_average_rates
+
+        avg_rates = ExchangeRateCurrencyRate.by_type(ExchangeRateCurrencyRate::AVERAGE_RATE_TYPE)
+
+        expect(avg_rates.all? { |r| r.validity_start_date == Time.zone.today.beginning_of_month }).to be true
+        expect(avg_rates.all? { |r| r.validity_end_date == Time.zone.today.end_of_month }).to be true
+
+        expect(avg_rates.by_currency('USD').first.rate).to eq(1.2264056116916666)
+        expect(avg_rates.by_currency('EUR').first.rate).to eq(1.1491571170166666)
+        expect(avg_rates.by_currency('AUD').first.rate).to eq(1.8997111260800001)
+        expect(avg_rates.by_currency('AED').first.rate).to eq(4.3747)
+        expect(avg_rates.by_currency('KZT').first.rate).to eq(545.0922)
+        expect(TariffSynchronizer::FileService).to have_received(:write_file).with(expected_filepath, expected_csv)
+      end
+    end
+
+    context 'when run on a valid date' do
+      let(:test_date) { Date.new(2023, 12, 31) }
+
+      after do
+        travel_back
+      end
+
+      it_behaves_like 'creating the average rates'
+    end
+
+    context 'when run on an invalid date' do
+      let(:test_date) { Date.new(2023, 10, 25) }
+
+      after do
+        travel_back
+      end
+
+      it 'will return an argument error' do
+        expect { create_average_rates }.to raise_error(ArgumentError)
+      end
+
+      context 'with force run enabled' do
+        let(:force_run) { true }
+
+        it_behaves_like 'creating the average rates'
+      end
+    end
+  end
+
+  def setup_data
```

Comment:
I know lots of setup data but this is to ensure all test cases are covered, it could be done with a massive sql insert query but i think this is clearer in what objects are being created

Reply:
Plus this is like an integration test for ensuring all the correct data is created

Reply:
yeah, up to you Sam, have a look if we can use just the bare minimum factories,
the more we use, the slower are the tests.